spark/v3.0/spark/src/main/java/org/apache/iceberg/spark/SparkSessionCatalog.java

```diff
@@ -123,11 +123,10 @@ public Table loadTable(Identifier ident) throws NoSuchTableException {
 
   @Override
   public void invalidateTable(Identifier ident) {
-    if (icebergCatalog.tableExists(ident)) {
-      icebergCatalog.invalidateTable(ident);
-    } else {
-      getSessionCatalog().invalidateTable(ident);
-    }
+    // We do not need to check whether the table exists and whether
+    // it is an Iceberg table to reduce remote service requests.
+    icebergCatalog.invalidateTable(ident);
```

Comment:
So we always have to do at least one catalog load here. Since we have a caching catalog this is at most a single remote lookup but in most circumstances should probably be zero because the reference is cached.

The code for the SparkCatalog 'invalidatetable' method starts by doing a 'load' operation, so either that call will use a cached reference or do a remote lookup. If the 'exists' call was remote than it the table reference would be cached and the invalidate call would use that cache. If the exists call wasn't remote it also would be cached that means the reference was cached already and invalidate should also use that cache. So the number of remote lookups should be at most one both with this patch and without.

This change now also always calls the underlying session catalogs invalidate method and I'm not sure what the consequences of that so I would probably recommend leaving this as is unless we have some evidence this is causing an issue.

I think if we want to remove the remote calls we have to first go to the invalidateTables method and do some mods there.

First, check whether the catalog is caching or not. For non caching or no cache timeout catalogs we can return a noop. For tables backed by a catalog with a cache we would probably need to add a dropCache function to the Iceberg cachingcatalog api or something like that so we don't have to rely on calling load.

Reply:
@RussellSpitzer Sorry, I forgot link to https://github.com/apache/iceberg/pull/3837. In https://github.com/apache/iceberg/pull/3837, we add a `invalidateTable` method to `Catalog` interface and remove `load` calls in `SparkCatalog.invalidateTable`.

Reply:
> This change now also always calls the underlying session catalogs invalidate method and I'm not sure what the consequences ...

For this, I have no good idea at present.
Can we rely on the conventions made in Spark for the 'invalidetable' interface document?

Reply:
What about updating `invalidateTable` to return `true` if a table was removed from the cache? Then we could use that to avoid calling `getSessionCatalog().invalidateTable(ident)` at least when there was a table loaded.

Reply:
I think I'm just being paranoid so I'm ok with calling both ... I think it should be fine as long as no one breaks contracts

Reply:
> What about updating invalidateTable to return true if a table was removed from the cache?

I have updated `Catalog.invalidateTable` to return whether the table is cached. However, due to Spark's `TableCatalog.invalidateTable` has no return value, I have to add `invalidateTableIfCached` method to `BaseCatalog` to workaround it.